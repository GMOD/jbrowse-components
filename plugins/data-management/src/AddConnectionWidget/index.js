import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'

export { default as ReactComponent } from './components/AddConnectionWidget'
export { default as stateModel } from './model'
export const configSchema = ConfigurationSchema('AddConnectionWidget', {})
