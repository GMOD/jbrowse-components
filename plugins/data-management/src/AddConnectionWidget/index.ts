import { lazy } from 'react'
import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { WidgetType } from '@jbrowse/core/pluggableElementTypes'

import stateModel from './model'
import type PluginManager from '@jbrowse/core/PluginManager'
const configSchema = ConfigurationSchema('AddConnectionWidget', {})

export default function AddConnectionWidgetF(pluginManager: PluginManager) {
  pluginManager.addWidgetType(() => {
    return new WidgetType({
      name: 'AddConnectionWidget',
      heading: 'Add a connection',
      configSchema,
      stateModel,
      ReactComponent: lazy(() => import('./components/AddConnectionWidget')),
    })
  })
}
