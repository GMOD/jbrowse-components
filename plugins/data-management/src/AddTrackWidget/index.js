import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'

export { default as ReactComponent } from './components/AddTrackWidget'
export { default as stateModelFactory } from './model'
export const configSchema = ConfigurationSchema('AddTrackWidget', {})
