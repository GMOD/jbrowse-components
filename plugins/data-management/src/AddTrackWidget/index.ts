import { ConfigurationSchema } from '@jbrowse/core/configuration'

export { default as stateModelFactory } from './model'
export const configSchema = ConfigurationSchema('AddTrackWidget', {})
