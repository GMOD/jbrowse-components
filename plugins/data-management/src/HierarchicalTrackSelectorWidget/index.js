import { ConfigurationSchema } from '@jbrowse/core/configuration'

export { default as ReactComponent } from './components/HierarchicalTrackSelector'
export { default as stateModelFactory } from './model'
export const configSchema = ConfigurationSchema(
  'HierarchicalTrackSelectorWidget',
  {},
)
