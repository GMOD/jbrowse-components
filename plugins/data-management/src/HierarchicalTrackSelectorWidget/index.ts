import stateModelFactory, { HierarchicalTrackSelectorModel } from './model'
import configSchema from './configSchema'

export { stateModelFactory, configSchema }
export type { HierarchicalTrackSelectorModel }
