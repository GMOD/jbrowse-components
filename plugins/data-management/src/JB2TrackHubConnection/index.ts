import { ConnectionType } from '@jbrowse/core/pluggableElementTypes'

import configSchema from './configSchema'
import stateModelFactory from './model'

import type PluginManager from '@jbrowse/core/PluginManager'

export default function JB2TrackHubConnectionF(pluginManager: PluginManager) {
  pluginManager.addConnectionType(() => {
    return new ConnectionType({
      name: 'JB2TrackHubConnection',
      configSchema,
      stateModel: stateModelFactory(pluginManager),
      displayName: 'JB2 Track Hub',
      description: 'A JBrowse 2 config file based trackhub',
      url: 'https://jbrowse.org/jb2/',
    })
  })
}
