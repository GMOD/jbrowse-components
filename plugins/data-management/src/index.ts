import ConnectionType from '@gmod/jbrowse-core/pluggableElementTypes/ConnectionType'
import WidgetType from '@gmod/jbrowse-core/pluggableElementTypes/WidgetType'
import Plugin from '@gmod/jbrowse-core/Plugin'
import PluginManager from '@gmod/jbrowse-core/PluginManager'
import {
  SessionWithWidgets,
  isAbstractMenuManager,
} from '@gmod/jbrowse-core/util'
import NoteAddIcon from '@material-ui/icons/NoteAdd'
import InputIcon from '@material-ui/icons/Input'
import {
  configSchema as ucscConfigSchema,
  modelFactory as ucscModelFactory,
} from './ucsc-trackhub'
import {
  ReactComponent as AddTrackReactComponent,
  stateModelFactory as AddTrackStateModelFactory,
  configSchema as AddTrackConfigSchema,
} from './AddTrackWidget'
import {
  ReactComponent as AddConnectionReactComponent,
  stateModel as AddConnectionStateModel,
  configSchema as AddConnectionConfigSchema,
} from './AddConnectionWidget'
import {
  ReactComponent as HierarchicalTrackSelectorReactComponent,
  stateModelFactory as HierarchicalTrackSelectorStateModelFactory,
  configSchema as HierarchicalTrackSelectorConfigSchema,
} from './HierarchicalTrackSelectorWidget'

export default class extends Plugin {
  name = 'DataManagementPlugin'

  install(pluginManager: PluginManager) {
    pluginManager.addConnectionType(
      () =>
        new ConnectionType({
          name: 'UCSCTrackHubConnection',
          configSchema: ucscConfigSchema,
          stateModel: ucscModelFactory(pluginManager),
          displayName: 'UCSC Track Hub',
          description: 'A track or assembly hub in the Track Hub format',
          url: '//genome.ucsc.edu/goldenPath/help/hgTrackHubHelp.html#Intro',
        }),
    )

    pluginManager.addWidgetType(() => {
      return new WidgetType({
        name: 'HierarchicalTrackSelectorWidget',
        heading: 'Available Tracks',
        configSchema: HierarchicalTrackSelectorConfigSchema,
        stateModel: HierarchicalTrackSelectorStateModelFactory(pluginManager),
        ReactComponent: HierarchicalTrackSelectorReactComponent,
      })
    })

    pluginManager.addWidgetType(() => {
      return new WidgetType({
        name: 'AddTrackWidget',
        heading: 'Add a track',
        configSchema: AddTrackConfigSchema,
        stateModel: AddTrackStateModelFactory(pluginManager),
        ReactComponent: AddTrackReactComponent,
      })
    })

    pluginManager.addWidgetType(() => {
      return new WidgetType({
        name: 'AddConnectionWidget',
        heading: 'Add a connection',
        configSchema: AddConnectionConfigSchema,
        stateModel: AddConnectionStateModel,
        ReactComponent: AddConnectionReactComponent,
      })
    })
  }

  configure(pluginManager: PluginManager) {
    if (isAbstractMenuManager(pluginManager.rootModel)) {
      pluginManager.rootModel.appendToMenu('File', {
        label: 'Open new track',
        icon: NoteAddIcon,
        onClick: (session: SessionWithWidgets) => {
          const widget = session.addWidget('AddTrackWidget', 'addTrackWidget')
          session.showWidget(widget)
        },
      })
      pluginManager.rootModel.appendToMenu('File', {
        label: 'Open new connection',
        icon: InputIcon,
        onClick: (session: SessionWithWidgets) => {
          const widget = session.addWidget(
            'AddConnectionWidget',
            'addConnectionWidget',
          )
          session.showWidget(widget)
        },
      })
    }
  }
}
