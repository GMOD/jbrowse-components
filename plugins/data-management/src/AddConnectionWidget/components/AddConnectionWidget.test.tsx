import React from 'react'
import { ThemeProvider } from '@emotion/react'
import { createJBrowseTheme } from '@jbrowse/core/ui'
import { createTestSession } from '@jbrowse/web/src/rootModel'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

// locals
import AddConnectionWidget2 from './AddConnectionWidget'

jest.mock('@jbrowse/web/src/makeWorkerInstance', () => () => {})

function AddConnectionWidget({ model }: { model: unknown }) {
  return (
    <ThemeProvider theme={createJBrowseTheme()}>
      <AddConnectionWidget2 model={model} />
    </ThemeProvider>
  )
}

function makeSession() {
  const session = createTestSession()
  session.addAssemblyConf({
    name: 'volMyt1',
    sequence: {
      trackId: 'ref0',
      type: 'ReferenceSequenceTrack',
      adapter: {
        type: 'FromConfigSequenceAdapter',
        features: [
          {
            refName: 'ctgA',
            uniqueId: 'firstId',
            start: 0,
            end: 1000,
            seq: 'cattgttgcggagttgaacaACGGCATTAGGAACACTTCCGTCTCtcacttttatacgattatgattggttctttagccttggtttagattggtagtagtagcggcgctaatgctacctgaattgagaactcgagcgggggctaggcaaattctgattcagcctgacttctcttggaaccctgcccataaatcaaagggttagtgcggccaaaacgttggacaacggtattagaagaccaacctgaccaccaaaccgtcaattaaccggtatcttctcggaaacggcggttctctcctagatagcgatctgtggtctcaccatgcaatttaaacaggtgagtaaagattgctacaaatacgagactagctgtcaccagatgctgttcatctgttggctccttggtcgctccgttgtacccaggctactttgaaagagcgcagaatacttagacggtatcgatcatggtagcatagcattctgataacatgtatggagttcgaacatccgtctggggccggacggtccgtttgaggttggttgatctgggtgatagtcagcaagatagacgttagataacaaattaaaggattttaccttagattgcgactagtacaacggtacatcggtgattcgcgctctactagatcacgctatgggtaccataaacaaacggtggaccttctcaagctggttgacgcctcagcaacataggcttcctcctccacgcatctcagcataaaaggcttataaactgcttctttgtgccagagcaactcaattaagcccttggtaccgtgggcacgcattctgtcacggtgaccaactgttcatcctgaatcgccgaatgggactatttggtacaggaatcaagcggatggcactactgcagcttatttacgacggtattcttaaagtttttaagacaatgtatttcatgggtagttcggtttgttttattgctacacaggctcttgtagacgacctacttagcactacgg',
          },
        ],
      },
    },
  })
  const model = session.addWidget('AddConnectionWidget', 'addConnectionWidget')
  return { model, session }
}

function renderWidget() {
  const user = userEvent.setup()
  const { model, session } = makeSession()
  const result = render(<AddConnectionWidget model={model} />)
  return { ...result, session, model, user }
}

test('renders', () => {
  const { container } = renderWidget()
  expect(container).toMatchSnapshot()
}, 20000)

test('can handle a custom UCSC trackHub URL', async () => {
  jest
    .spyOn(global, 'fetch')
    .mockImplementation(async (url: string | Request | URL) => {
      const urlText = `${url}`
      if (urlText.endsWith('hub.txt')) {
        return new Response(`hub TestHub
shortLabel Test Hub
longLabel Test Genome Informatics Hub for human DNase and RNAseq data
genomesFile genomes.txt
email genome@test.com
descriptionUrl test.html
`)
      } else if (urlText.endsWith('genomes.txt')) {
        return new Response(`genome volMyt1
trackDb hg19/trackDb.txt
`)
      } else if (urlText.endsWith('trackDb.txt')) {
        return new Response(`track dnaseSignal
bigDataUrl dnaseSignal.bigWig
shortLabel DNAse Signal
longLabel Depth of alignments of DNAse reads
type bigWig
`)
      }
      throw new Error('unknown')
    })

  const {
    session,
    user,
    findByText,
    getAllByRole,
    findAllByText,
    findByDisplayValue,
  } = renderWidget()
  expect(session.connections.length).toBe(0)
  await user.click(getAllByRole('combobox')[0]!)
  const ucscTrackHubSelection = await findAllByText('UCSC Track Hub')
  await user.click(ucscTrackHubSelection.at(-1)!)
  await user.click(await findByText('Next'))
  await user.type(
    await findByDisplayValue('nameOfConnection'),
    'Test UCSC connection name',
  )
  await user.type(
    await findByDisplayValue('http://mysite.com/path/to/hub.txt'),
    'http://test.com/hub.txt',
  )
  await user.click(await findByText('Connect'))
  expect(session.sessionConnections.length).toBe(1)
}, 20000)

test('can handle a custom JBrowse 1 data directory URL', async () => {
  jest
    .spyOn(global, 'fetch')
    .mockImplementation(async (url: RequestInfo | URL) => {
      const urlText = `${url}`
      let responseText = ''
      if (urlText.endsWith('trackList.json')) {
        responseText = '{}'
      } else if (urlText.endsWith('refSeqs.json')) {
        responseText = '[]'
      }
      return new Response(responseText)
    })

  const {
    findByTestId,
    getAllByRole,
    findByText,
    findByPlaceholderText,
    findByDisplayValue,
    user,
    session,
  } = renderWidget()
  expect(session.connections.length).toBe(0)
  await user.click(getAllByRole('combobox')[0]!)
  await user.click(await findByText('JBrowse 1 Data'))
  await user.click(await findByText('Next'))
  await user.type(await findByDisplayValue('nameOfConnection'), 'testing')
  await user.type(
    await findByDisplayValue('http://mysite.com/jbrowse/data/'),
    'http://test.com/jbrowse/data/',
  )
  await user.click(await findByText('Add item'))
  await user.type(await findByPlaceholderText('add new'), 'volMyt1')
  await user.click(await findByTestId('stringArrayAdd-assemblyNames'))
  await user.click(await findByText('Connect'))
  expect(session.sessionConnections.length).toBe(1)
}, 20000)
