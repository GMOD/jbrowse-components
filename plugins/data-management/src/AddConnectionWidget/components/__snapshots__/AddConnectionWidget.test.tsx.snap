// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <div
    class="css-3o7gsw-root"
  >
    <div
      class="MuiStepper-root MuiStepper-vertical css-nzjsqw-MuiStepper-root-stepper"
    >
      <div
        class="MuiStep-root MuiStep-vertical css-enonti-MuiStep-root"
      >
        <span
          class="MuiStepLabel-root MuiStepLabel-vertical css-122jq7f-MuiStepLabel-root"
        >
          <span
            class="MuiStepLabel-iconContainer Mui-active css-1b2ju2j-MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root Mui-active css-1jm3coa-MuiSvgIcon-root-MuiStepIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                class="MuiStepIcon-text css-1i1npeq-MuiStepIcon-text"
                dominant-baseline="central"
                text-anchor="middle"
                x="12"
                y="12"
              >
                1
              </text>
            </svg>
          </span>
          <span
            class="MuiStepLabel-labelContainer css-p8xe3-MuiStepLabel-labelContainer"
          >
            <span
              class="MuiStepLabel-label Mui-active css-vxmcgq-MuiStepLabel-label"
            >
              Select a Connection Type
            </span>
          </span>
        </span>
        <div
          class="MuiStepContent-root css-69uykz-MuiStepContent-root"
        >
          <div
            class="MuiCollapse-root MuiCollapse-vertical MuiStepContent-transition MuiCollapse-entered css-scib18-MuiCollapse-root-MuiStepContent-transition"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper MuiCollapse-vertical css-1x6hinx-MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner MuiCollapse-vertical css-1i4ywhz-MuiCollapse-wrapperInner"
              >
                <form
                  autocomplete="off"
                >
                  <div
                    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
                  >
                    <label
                      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-cgj0jf-MuiFormLabel-root-MuiInputLabel-root"
                      data-shrink="true"
                      for=":r0:"
                      id=":r0:-label"
                    >
                      connectionType
                    </label>
                    <div
                      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-b32mwt-MuiInputBase-root-MuiOutlinedInput-root"
                    >
                      <div
                        aria-controls=":r1:"
                        aria-describedby=":r0:-helper-text"
                        aria-expanded="false"
                        aria-haspopup="listbox"
                        aria-labelledby=":r0:-label :r0:"
                        class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-w76bbz-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                        id=":r0:"
                        role="combobox"
                        tabindex="0"
                      >
                        UCSC Track Hub
                      </div>
                      <input
                        aria-hidden="true"
                        aria-invalid="false"
                        class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
                        tabindex="-1"
                        value="UCSCTrackHubConnection"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-e00l0k-MuiSvgIcon-root-MuiSelect-icon"
                        data-testid="ArrowDropDownIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                      <fieldset
                        aria-hidden="true"
                        class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                      >
                        <legend
                          class="css-w1u3ce"
                        >
                          <span>
                            connectionType
                          </span>
                        </legend>
                      </fieldset>
                    </div>
                    <p
                      class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled css-nxe85d-MuiFormHelperText-root"
                      id=":r0:-helper-text"
                    >
                      A track or assembly hub in the Track Hub format
                      <a
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1dmujzy-MuiButtonBase-root-MuiIconButton-root"
                        href="//genome.ucsc.edu/goldenPath/help/hgTrackHubHelp.html#Intro"
                        rel="noopener noreferrer"
                        tabindex="0"
                        target="_blank"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-14500qs-MuiSvgIcon-root"
                          data-testid="OpenInNewIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3z"
                          />
                        </svg>
                      </a>
                    </p>
                  </div>
                </form>
                <div
                  class="css-14dqgmq-actionsContainer"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-6g0gn4-MuiButtonBase-root-MuiButton-root-button"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    Back
                  </button>
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-14hk3uo-MuiButtonBase-root-MuiButton-root-button"
                    data-testid="addConnectionNext"
                    tabindex="0"
                    type="button"
                  >
                    Next
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiStepConnector-root MuiStepConnector-vertical Mui-disabled css-1afu6qk-MuiStepConnector-root"
      >
        <span
          class="MuiStepConnector-line MuiStepConnector-lineVertical css-1deaddp-MuiStepConnector-line"
        />
      </div>
      <div
        class="MuiStep-root MuiStep-vertical css-enonti-MuiStep-root"
      >
        <span
          class="MuiStepLabel-root MuiStepLabel-vertical Mui-disabled css-122jq7f-MuiStepLabel-root"
        >
          <span
            class="MuiStepLabel-iconContainer Mui-disabled css-1b2ju2j-MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root css-1jm3coa-MuiSvgIcon-root-MuiStepIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                class="MuiStepIcon-text css-1i1npeq-MuiStepIcon-text"
                dominant-baseline="central"
                text-anchor="middle"
                x="12"
                y="12"
              >
                2
              </text>
            </svg>
          </span>
          <span
            class="MuiStepLabel-labelContainer css-p8xe3-MuiStepLabel-labelContainer"
          >
            <span
              class="MuiStepLabel-label Mui-disabled css-vxmcgq-MuiStepLabel-label"
            >
              Configure Connection
            </span>
          </span>
        </span>
        <div
          class="MuiStepContent-root MuiStepContent-last css-1r4r9eo-MuiStepContent-root"
        />
      </div>
    </div>
  </div>
</div>
`;
